editor/components/block-drop-zone/index.js

```diff
@@ -132,16 +142,15 @@ export default compose(
 				updateBlockAttributes( ...args );
 			},
 			moveBlockToPosition( uid, fromRootUID, index ) {
-				const { rootUID, layout } = ownProps;
-				moveBlockToPosition( uid, fromRootUID, rootUID, layout, index );
+				const { canInsertBlockType, getBlockName, rootUID, layout } = ownProps;
+				const blockName = getBlockName( uid );
+				// currently the only constraint to move inside the same parent is locking
+				// locking was already handled
+				// it is not possible to use drag & drop if locking is active
+				if ( rootUID === fromRootUID || canInsertBlockType( blockName, rootUID ) ) {
```

Comment:
While this is good for now, I wonder if we should bake a validation mechanism inside `DropZone`. That way we can change the cursor to `not-allowed` when we're hovering an area where we can't drop.